Add support for the Blockstream Jade hww

JamieDriver opened the pull request:
Plugin for Blockstream Jade hww - following pattern of other hww plugins.

SomberNight commented:
Hi! Thanks for this.
Do you work for Jade/Blockstream? If so, would it be possible to send us two devices for testing purposes? We can discuss details in email.

JamieDriver commented:
Hi, 
Yes indeed.  I'll raise your question with the mgmt team.  Cheers.

evliu commented:
Thanks for this! Can't get my new Jade to work with Spectrum so looking for other alternatives.

dangvietquang commented:
Ladies and Gentlemen. Hacker attacks for thousands of years, still can't be.
If not grasp the key structure of the security system. That is 
understandable that the investigating agency always takes as a basis. 
Whether it was necessary or not, this case was unacceptable. Although clever
camouflage takes many forms. For the purpose of removing the system's 
identity with the founder. Although removing all that identification, of the
system with the founder. Create new barriers to erase the truth. Then at 
this point, all countries have archived, the identification of the system 
many years, many months ago. Because this was a conspiracy, a plan many 
years ago. Due to being controlled from the beginning. This incident has 
been a special case since 2013 in places of international responsibility. 
Therefore, Humans can only be as honest as computer systems. That's a smart 
person. All mistakes will be exposed at the great Blockchain system. This is
what makes Bitcoin shine, not to be overshadowed. Whoever it is, what it 
does, and how sophisticated it is.
---------- Původní e-mail ----------
Od: Everest Liu ***@***.***>
Komu: spesmilo/electrum ***@***.***>
Datum: 3. 2. 2022 19:18:41
Předmět: Re: [spesmilo/electrum] Add support for the Blockstream Jade hww 
(PR #7633) 
"


Thanks for this! Can't get my new Jade to work with Spectrum so looking for 
other alternatives.

—
Reply to this email directly, view it on GitHub
(https://github.com/spesmilo/electrum/pull/7633#issuecomment-1029269322), or
unsubscribe
(https://github.com/notifications/unsubscribe-auth/AWARNNLD2OSVPIRPXFOLUJLUZLBHTANCNFSM5MJIYJDQ)
.
Triage notifications on the go with GitHub Mobile for iOS
(https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675)
or Android
(https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub)
. 
You are receiving this because you are subscribed to this thread. Message 
ID: ***@***.***>
"

evliu commented:
@JamieDriver I've discovered with my newer Jade, pyserial 3.4 can't read it but pyserial 3.5 can; I've documented the issue here https://github.com/cryptoadvance/specter-desktop/issues/1480#issuecomment-1076837538

Any updates on this PR?

JamieDriver commented:
@evliu  - Thanks for testing it!

@SomberNight - Are you waiting for anything from me/Blockstream ?  Cheers.

SomberNight reviewed (changes requested):
Thanks for the PR. The code largely looks good.
See minor comments inline.

I've tested with a real device on Linux, both singlesig and multisig, and it works ok.

-----

One issue I have is that often there is no UI feedback to the user that something is going on.
For instance, try signing a message, or showing an address on the device screen, or signing a tx. The other plugins display a popup telling the user that there is something going on and they should interact with their hw device, e.g. "Confirm Transaction on your Ledger device...".
This is usually done using the `handler`, see https://github.com/spesmilo/electrum/blob/9e0a0af81a7dab4fe22aeb007fb5560f48dc6433/electrum/plugins/ledger/ledger.py#L561
but I guess it's not too critical - it could be done in a follow-up if you prefer

-----

There is a more important concern I have, which is that I cannot get this to work on Windows.
Originally, `pyserial` could not see that there is a device connected at all, but I've found the [blockstream docs](https://help.blockstream.com/hc/en-us/articles/900005443223-Nothing-happens-when-I-plug-my-Blockstream-Jade-to-my-computer) which instruct the user to install a driver. This should at least be mentioned somewhere. Or is there maybe a workaround? Could we bundle the driver with the application somehow?

Regardless, even after installing the driver, I have problems. Specifically, `pyserial` can now see the device:
```
>>> from serial.tools import list_ports; print(list_ports.comports())
[<serial.tools.list_ports_common.ListPortInfo object at 0x0000025811E5BCA0>]
```
however, when using Electrum, and scanning for hw devices in the wizard I either get a permission error, or the whole application hangs ("not responding") and I have to kill the process, or sometimes I get an EOFError().

```
I | plugin.DeviceMgr | scanning devices...
D | util.profiler | DeviceMgr.scan_devices 0.2303
E | plugin.DeviceMgr | failed to create client for jade at COM4: SerialException("could not open port 'COM4': PermissionError(13, 'Access is denied.', None, 5)")
W | gui.qt.installwizard.InstallWizard | error getting device infos for jade: could not open port 'COM4': PermissionError(13, 'Access is denied.', None, 5)
```
```
E | plugin.DeviceMgr | failed to create client for jade at COM4: EOFError()
```

Could you please test with Windows?

-----

note: testing with fw version 0.1.33 (no-radio variant).
I have not tested on macOS.

JamieDriver commented on electrum/plugins/jade/jade.py:
```diff
@@ -0,0 +1,458 @@
+    SUPPORTED_XTYPES = ('standard', 'p2wpkh-p2sh', 'p2wpkh', 'p2wsh-p2sh', 'p2wsh')
+    MIN_SUPPORTED_FW_VERSION = semver.VersionInfo(0, 1, 32)
+
+    # For testing with qemu simulator
+    SIMULATOR_PATH = 'tcp:127.0.0.1:2222'
+    SIMULATOR_TEST_SEED = bytes.fromhex('b90e532426d0dc20fffe01037048c018e940300038b165c211915c672e07762c')
```
Agree - will disable as suggested (is of limited value in any case).

JamieDriver commented on electrum/plugins/jade/jade.py:
```diff
@@ -0,0 +1,458 @@
+
+    plugin: 'JadePlugin'
+
+    def __init__(self, d):
+        Hardware_KeyStore.__init__(self, d)
+        # Errors and other user interaction is done through the wallet's
+        # handler.  The handler is per-window and preserved across
+        # device reconnects
+        self.force_watching_only = False
+        self.cfg = d.get('cfg', {'mode': 0})
+
+    def dump(self):
+        obj = Hardware_KeyStore.dump(self)
+        obj['cfg'] = self.cfg
+        return obj
```
Apologies - yes I started by looking at some of the other hwws and making a 'matching' implementation ... sorry if pulled across unnecessary snippets.  
Will remove these functions and let the base class handle it.

JamieDriver commented on electrum/plugins/jade/jade.py:
```diff
@@ -0,0 +1,458 @@
+import os
+import base64
+import semver
```
No problem.  In some of our code we use 'semver' as it handles eg. beta and rc versions - but yes since we don't really expect those versions to be in use out in the wild the more straightforward check you suggest should not cause any issues - and as you say it removes a dependency.

JamieDriver commented on electrum/plugins/jade/jade.py:
```diff
@@ -0,0 +1,458 @@
+        raise UserFacingException(_('Encryption and decryption are not implemented by {}').format(self.device))
+
+    @runs_in_hwd_thread
+    def sign_message(self, sequence, message, password):
```
Thanks - will rebase & update.

JamieDriver commented:
FYI:
I have been unable to build the appimage using: `contrib/build-linux/build.sh`, even on 'master'.
I get:
`E: Version '1.12.2-1ubuntu1.2' for 'libdbus-1-3' was not found` from the apt-install step in the Dockerfile.

After a quick eyeball of:  `https://packages.ubuntu.com/search?keywords=libdbus-1-3` I locally changed it to  '1.12.2-1ubuntu1.3' and that seems to work.
I've not submitted that in this PR since it's unrelated to Jade - I just thought I'd let you know.
(Or am happy to raise it in its own PR if so desired.)

SomberNight commented on electrum/plugins/jade/jade.py:
```diff
@@ -0,0 +1,464 @@
+    def authenticate(self):
+        # Ensure Jade unlocked - always call hw unit at least once
+        # If the hw is already unlocked, this call returns immediately/no-op
+        authenticated = False
+        while not authenticated:
+            authenticated = self.jade.auth_user(self._network())
```
It would be nice if we could show a message on screen to the user when they have to enter the PIN on the device, e.g. with `self.handler.show_message` as in the other user-interactions-prompts. I realise that `handler.show_message`/`handler.finished` does not compose with itself however, so this is not easy to do with how the code is currently written.
I guess it's fine as-is :)

SomberNight reviewed (commented):
Thanks for the quick changes. It looks good.

Ping back if you had a chance to test on Windows. Feel free to just use a VM and pass through the USB device, I do that all the time (but in this specific case I have issues even on a real windows machine).

SomberNight commented on electrum/wallet_db.py:
```diff
@@ -129,7 +129,7 @@ def get_split_accounts(self):
             result = [data1, data2]
 
-        elif wallet_type in ['bip44', 'trezor', 'keepkey', 'ledger', 'btchip', 'digitalbitbox', 'safe_t']:
+        elif wallet_type in ['bip44', 'trezor', 'keepkey', 'ledger', 'btchip', 'digitalbitbox', 'safe_t', 'jade']:
```
this is not needed, see https://github.com/spesmilo/electrum/commit/cbc69742b958aaacdad9612d7cf1564c32205efb

JamieDriver replied:
Removed.

SomberNight commented on electrum/wallet_db.py:
```diff
@@ -258,7 +258,7 @@ def _convert_wallet_type(self):
             self.put('keystore', d)
 
-        elif wallet_type in ['trezor', 'keepkey', 'ledger', 'digitalbitbox', 'safe_t']:
+        elif wallet_type in ['trezor', 'keepkey', 'ledger', 'digitalbitbox', 'safe_t', 'jade']:
```
same as above

JamieDriver replied:
Removed.

SomberNight commented on electrum/plugins/jade/jade.py:
```diff
@@ -0,0 +1,464 @@
+class JadePlugin(HW_PluginBase):
+    keystore_class = Jade_KeyStore
+    minimum_library = (0, 0, 1)
+    client = None
```
this line is copied dead code, see https://github.com/spesmilo/electrum/commit/aab8e664ed88a1f9c63e9a3d409e63775020626d

JamieDriver replied:
Agreed, removed.

SomberNight commented on electrum/plugins/jade/jade.py:
```diff
@@ -0,0 +1,464 @@
+
+    @runs_in_hwd_thread
+    def create_client(self, device, handler):
+        if handler:
+            self.handler = handler
```
this is copied-over dead code, see https://github.com/spesmilo/electrum/commit/30623c352907d80dab421cbfb3dabf9c3847ac57

JamieDriver replied:
Cheers, removed.

SomberNight reviewed (commented):
some more trivial things I noticed

SomberNight reviewed (commented):
Thanks for the changes, looking good.
(only waiting re Windows issues now)

JamieDriver commented:
Cheers.  We'll def get some testing/investigation in over the next few working days and get back to you.  
Many thanks for your time on this.

JamieDriver commented:
@SomberNight 
Sorry ... Have I missed something related to the windows build ?  When we download the windows CI artifact and try to run it we get errors about not finding the jade lib.
```
jade: (error getting device infos)
    Missing libraries for jade.
    Make sure you install it with python3
```
I assume they should all be bundled into the exe (much like the linux AppImage) so I am assuming I've missed some dependencies or manifest type file ?
Many thanks.

SomberNight commented:
>  Sorry ... Have I missed something related to the windows build ? When we download the windows CI artifact and try to run it we get errors about not finding the jade lib.

Ah, indeed, the binary is missing some dependencies... Try with this patch:
```patch
diff --git a/contrib/build-wine/deterministic.spec b/contrib/build-wine/deterministic.spec
index f939964cf0..969897ab5d 100644
--- a/contrib/build-wine/deterministic.spec
+++ b/contrib/build-wine/deterministic.spec
@@ -22,6 +22,8 @@ hiddenimports += collect_submodules('keepkeylib')
 hiddenimports += collect_submodules('websocket')
 hiddenimports += collect_submodules('ckcc')
 hiddenimports += collect_submodules('bitbox02')
+hiddenimports += ['electrum.plugins.jade.jade']
+hiddenimports += ['electrum.plugins.jade.jadepy.jade']
 hiddenimports += ['PyQt5.QtPrintSupport']  # needed by Revealer
 
 
diff --git a/contrib/osx/osx.spec b/contrib/osx/osx.spec
index 1208bea166..3e9fe0148f 100644
--- a/contrib/osx/osx.spec
+++ b/contrib/osx/osx.spec
@@ -30,6 +30,8 @@ hiddenimports += collect_submodules('keepkeylib')
 hiddenimports += collect_submodules('websocket')
 hiddenimports += collect_submodules('ckcc')
 hiddenimports += collect_submodules('bitbox02')
+hiddenimports += ['electrum.plugins.jade.jade']
+hiddenimports += ['electrum.plugins.jade.jadepy.jade']
 hiddenimports += ['PyQt5.QtPrintSupport']  # needed by Revealer
 
 datas = [
```

JamieDriver commented:
Many thanks for the suggested patch.  
After applying that, we grabbed the windows build artifacts and everything seemed to test successfully.  We did not appear to have any issues connecting to Jade, signing tx, etc.  We ran through the same test plan as we had for linux without issue.   
(I'll also get it run under MacOS asap).

The only way we could reproduce the 'Access is denied.' error, is when some other instance or app was already connected to that COM port.

Looking through the code I did spot one small 'opportunity for improvement', adding an explicit `disconnect()` call - didn't seem to be required (afaics) but have added it anyway as it feels better/cleaner ...

SomberNight commented:
> we grabbed the windows build artifacts and everything seemed to test successfully. We did not appear to have any issues connecting to Jade, signing tx, etc. We ran through the same test plan as we had for linux without issue.
> The only way we could reproduce the 'Access is denied.' error, is when some other instance or app was already connected to that COM port.

Weird. I cannot get it to work on Windows. I've tried on a real machine, and in a VM, both running from source and with a built exe. I've tried with multiple different USB ports, two different USB cables, two different physical jade hw devices.

I've been messing around with it for quite a bit, and seeing exceptions, the device not responding (and some calls blocking - typically the first call after `jade.connect()`, either `jade.get_version_info()` or `jade.add_entropy()`), or the device unexpectedly rebooting)...
Out of several dozens of attempts, I've managed to connect to the device properly and finish creating a wallet file only a single time - so it did work once - but I could not reproduce again.

What did you do to get it to work?

-----

Let me detail a scenario I've just tested, on a real machine running Windows 10 21H1 (10.0.19043.1706):

1. jade turned off, usb unplugged
2. plug jade via usb, wait until splash screen finishes
    - device screen says "Connect jade to a Blockstream Green companion app"
3. start electrum
4. open wizard in electrum (might already open to that, or menubar>File>New/Restore
5. create new wallet. "Standard wallet">"Use a hardware device", press next
6. wizard shows "Debug message\njade: (error getting device infos)"
  meanwhile, jade seems to have *restarted* (screen flashed, splash screen appeared again, splash screen finished)
  exc:
	```
	E | plugin.DeviceMgr | failed to create client for jade at COM5: EOFError()
	Traceback (most recent call last):
	  File "...\electrum\electrum\plugin.py", line 601, in unpaired_device_infos
		client = self.create_client(device, handler, plugin)
	  File "...\electrum\electrum\plugin.py", line 362, in wrapper
		return run_in_hwd_thread(partial(func, *args, **kwargs))
	  File "...\electrum\electrum\plugin.py", line 355, in run_in_hwd_thread
		return fut.result()
	  File "...\Python310\lib\concurrent\futures\_base.py", line 446, in result
		return self.__get_result()
	  File "...\Python310\lib\concurrent\futures\_base.py", line 391, in __get_result
		raise self._exception
	  File "...\Python310\lib\concurrent\futures\thread.py", line 58, in run
		result = self.fn(*self.args, **self.kwargs)
	  File "...\electrum\electrum\plugin.py", line 452, in create_client
		client = plugin.create_client(device, handler)
	  File "...\electrum\electrum\plugin.py", line 362, in wrapper
		return run_in_hwd_thread(partial(func, *args, **kwargs))
	  File "...\electrum\electrum\plugin.py", line 352, in run_in_hwd_thread
		return func()
	  File "...\electrum\electrum\plugins\jade\jade.py", line 410, in create_client
		client = Jade_Client(device.path, plugin=self)
	  File "...\electrum\electrum\plugins\jade\jade.py", line 102, in __init__
		verinfo = self.jade.get_version_info()
	  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 185, in get_version_info
		return self._jadeRpc('get_version_info')
	  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 159, in _jadeRpc
		reply = self.jade.make_rpc_call(request, long_timeout)
	  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 653, in make_rpc_call
		reply = self.read_response(long_timeout)
	  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 632, in read_response
		return self.read_cbor_message()
	  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 601, in read_cbor_message
		message = cbor.load(self)
	  File "...\Python310\site-packages\cbor\cbor.py", line 271, in load
		return _loads(fp)[0]
	  File "...\Python310\site-packages\cbor\cbor.py", line 379, in _loads
		tb = _read_byte(fp)
	  File "...\Python310\site-packages\cbor\cbor.py", line 310, in _read_byte
		raise EOFError()
	EOFError
	W | gui.qt.installwizard.InstallWizard | error getting device infos for jade: 
	```
7. click "Next" in wizard again to trigger a rescan
	exc:
	```
	E | plugin.DeviceMgr | failed to create client for jade at COM5: SerialException("could not open port 'COM5': PermissionError(13, 'Access is denied.', None, 5)")
	Traceback (most recent call last):
	  File "...\electrum\electrum\plugin.py", line 601, in unpaired_device_infos
		client = self.create_client(device, handler, plugin)
	  File "...\electrum\electrum\plugin.py", line 362, in wrapper
		return run_in_hwd_thread(partial(func, *args, **kwargs))
	  File "...\electrum\electrum\plugin.py", line 355, in run_in_hwd_thread
		return fut.result()
	  File "...\Python310\lib\concurrent\futures\_base.py", line 446, in result
		return self.__get_result()
	  File "...\Python310\lib\concurrent\futures\_base.py", line 391, in __get_result
		raise self._exception
	  File "...\Python310\lib\concurrent\futures\thread.py", line 58, in run
		result = self.fn(*self.args, **self.kwargs)
	  File "...\electrum\electrum\plugin.py", line 452, in create_client
		client = plugin.create_client(device, handler)
	  File "...\electrum\electrum\plugin.py", line 362, in wrapper
		return run_in_hwd_thread(partial(func, *args, **kwargs))
	  File "...\electrum\electrum\plugin.py", line 352, in run_in_hwd_thread
		return func()
	  File "...\electrum\electrum\plugins\jade\jade.py", line 410, in create_client
		client = Jade_Client(device.path, plugin=self)
	  File "...\electrum\electrum\plugins\jade\jade.py", line 100, in __init__
		self.jade.connect()
	  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 136, in connect
		self.jade.connect()
	  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 529, in connect
		self.impl.connect()
	  File "...\electrum\electrum\plugins\jade\jadepy\jade_serial.py", line 32, in connect
		self.ser = serial.Serial(self.device, self.baud,
	  File "...\Python310\site-packages\serial\serialwin32.py", line 33, in __init__
		super(Serial, self).__init__(*args, **kwargs)
	  File "...\Python310\site-packages\serial\serialutil.py", line 244, in __init__
		self.open()
	  File "...\Python310\site-packages\serial\serialwin32.py", line 64, in open
		raise SerialException("could not open port {!r}: {!r}".format(self.portstr, ctypes.WinError()))
	serial.serialutil.SerialException: could not open port 'COM5': PermissionError(13, 'Access is denied.', None, 5)
	W | gui.qt.installwizard.InstallWizard | error getting device infos for jade: could not open port 'COM5': PermissionError(13, 'Access is denied.', None, 5)
	```
8. click "Next" in wizard again to trigger a rescan -- same as step7 ("could not open port 'COM5': PermissionError")
9. closed Electrum fully
	- during process shutdown, jade seems to have restarted again (screen flashed, splash screen appeared again, splash screen finished)
10. to be sure, unplugged jade from usb, and powered it off by long-pressing the green button

Then I plugged in the jade via usb again, waited until splash screen finishes, started electrum, and the same steps and errors reproduced from above, again.

-----

I wanted to see if can get the hw device to work on Windows using some other program, so I have also tried running the `green_qt` app (1.0.7) in a Windows VM, using the [binary from github](https://github.com/Blockstream/green_qt/releases/download/release_1.0.7/BlockstreamGreen_Windows_x86_64.zip).
After installing it, when I launch `Blockstream Green.exe`, an empty window appears for a few seconds and then the process exits (and the window disappears ofc).
This is all it logs:

```
[2022-05-17 17:10:42.418418] [app:info] Blockstream Green 1.0.7
[2022-05-17 17:10:42.418418] [app:info] System Information:
[2022-05-17 17:10:42.418418] [app:info]   Build ABI: x86_64-little_endian-llp64
[2022-05-17 17:10:42.418418] [app:info]   Build CPU Architecture: x86_64
[2022-05-17 17:10:42.418418] [app:info]   Current CPU Architecture: x86_64
[2022-05-17 17:10:42.418418] [app:info]   Kernel Type: winnt
[2022-05-17 17:10:42.418418] [app:info]   Kernel Version: 10.0.19044
[2022-05-17 17:10:42.418418] [app:info]   Product Type: windows
[2022-05-17 17:10:42.418418] [app:info]   Product Version: 10
[2022-05-17 17:10:42.418418] [app:info] Build Type: release
[2022-05-17 17:10:42.418418] [app:info] Data directory: "C:/Users/user/AppData/Local/Blockstream/Green"
[2022-05-17 17:10:42.418418] [app:info] Log file: "C:/Users/user/AppData/Local/Blockstream/Green/logs/1.0.7.txt"
[2022-05-17 17:10:43.092092] [app:debug] http: create session
[2022-05-17 17:10:43.107107] [gdk:info] net: thread started for gdk version aeab84f
[2022-05-17 17:10:43.107107] [gdk:info] change_state_to: requesting state connected
[2022-05-17 17:10:43.107107] [gdk:info] net: desired connected actual disconnected
[2022-05-17 17:10:43.107107] [gdk:info] net: connect to wss://prodwss.greenaddress.it/v2/ws
[2022-05-17 17:10:43.208208] [app:debug] http: waiting session to connect
[2022-05-17 17:10:43.455455] [gdk:info] net: connection successful
[2022-05-17 17:10:43.490490] [gdk:info] change_state_to: changed to connected
```
I guess this is most likely unrelated though -- I can open an issue upstream if you want.

JamieDriver commented:
That is odd as `green_qt` should def work - esp if you were not trying to connect to Jade/anything at that point.  

It also seems odd that Jade is crashing/rebooting when being sent messages.
Do you know what the Jade fw version is ?  It may say it on the bottom right of the screen ...

SomberNight commented:
>  Do you know what the Jade fw version is ? It may say it on the bottom right of the screen ...

It is 0.1.33, I've updated it prior, using green_qt on Linux.

JamieDriver commented:
One possibility re: `green_qt` not starting is that there is no (outdated?) open-gl support - if possible can you run with env var:
"QT_OPENGL=software".  
Cheers.

SomberNight commented:
> One possibility re: `green_qt` not starting is that there is no (outdated?) open-gl support - if possible can you run with env var: "QT_OPENGL=software".  

It did not help. No change in the log either. :/

SomberNight commented:
>> One possibility re: `green_qt` not starting is that there is no (outdated?) open-gl support - if possible can you run with env var: "QT_OPENGL=software".  
>
> It did not help. No change in the log either. :/

Oh, indeed the issue is graphics-related.
I am using VirtualBox 6.1.34, with "Graphics Controller" set to the default "VBoxSVGA", and "Enable 3D Acceleration" disabled.
Now I've tried enabling 3d acceleration, and green_qt no longer crashes inside the Windows guest.
However, there are now rendering bugs >.< I guess this is a bug of virtualbox - quick online searches of "virtualbox windows guest 3d acceleration" give multiple results detailing problems...


![Capture](https://user-images.githubusercontent.com/29142493/168883110-fc17706e-6be2-49f1-9fca-63584457367b.PNG)

SomberNight commented:
>> One possibility re: `green_qt` not starting is that there is no (outdated?) open-gl support - if possible can you run with env var: "QT_OPENGL=software".  
>
> It did not help. No change in the log either. :/

Ok, so the magic envvar to set is `QMLSCENE_DEVICE=softwarecontext`
`QT_OPENGL=software` has no seemingly effect and is not needed.
`green_qt` now works in the virtualbox windows guest without 3d acceleration, ~without visual glitches.
(Well, ok, there are still some minor glitches such as dropdown having transparent background, but anyway)

-----

`green_qt` running on the virtualbox windows guest can properly communicate with the jade hwd.
Electrum running on the same guest however exhibits the `EOFError`/`PermissionError`/hwd reboots issues, as above :/

JamieDriver commented:
Nice one!  Ok, so the Jade itself and cable etc. are ok.  
I guess it also means that the Silicon Labs driver must be ok also.  

If Jade is aborting/rebooting it sounds like it is getting some corruption/noise on the connection - or as if two things are trying to send it data at the same time, and blocks of bytes are getting interleaved (and hence look like junk/violate invariants etc.)

Will carry on looking at it tomorrow (I am tz utc+1) and trying to reproduce your experience.
Thanks for your assistance on this.

SomberNight commented:
Pasting this (snippet1) into a python interpreter makes the jade reboot:
```python
import os
from serial.tools import list_ports
from electrum.plugins.jade.jadepy.jade import JadeAPI

def enumerate_serial():
    paths = []
    for devinfo in list_ports.comports():
        if (devinfo.vid, devinfo.pid) in [(0x10c4, 0xea60), (0x1a86, 0x55d4)]:
            paths.append(devinfo.device)
    return paths

paths = enumerate_serial()
assert len(paths) == 1, paths
path = paths[0]

jade = JadeAPI.create_serial(path, timeout=1)
jade.connect()
jade.get_version_info()
jade.disconnect()
```

same for this (snippet2):
```python
import os
from serial.tools import list_ports
from electrum.plugins.jade.jadepy.jade import JadeAPI

def enumerate_serial():
    paths = []
    for devinfo in list_ports.comports():
        if (devinfo.vid, devinfo.pid) in [(0x10c4, 0xea60), (0x1a86, 0x55d4)]:
            paths.append(devinfo.device)
    return paths

paths = enumerate_serial()
assert len(paths) == 1, paths
path = paths[0]

jade = JadeAPI.create_serial(path, timeout=1)
jade.connect()
jade.get_version_info()
jade = JadeAPI.create_serial(path)
jade.connect()
```

-----

I see you are also experimenting with `Jade_Client.__init__`.
Note that if I anticipate when it is going to reboot and add sleep there, everything seems to work.

So e.g., inspired from snippet1, with this patch (on top of the original commit), electrum can use the jade:
```patch
diff --git a/electrum/plugins/jade/jade.py b/electrum/plugins/jade/jade.py
index 9e255e0b47..8440f9a994 100644
--- a/electrum/plugins/jade/jade.py
+++ b/electrum/plugins/jade/jade.py
@@ -1,6 +1,7 @@
 import os
 import base64
 from typing import Optional
+import time
 
 from electrum import bip32, constants
 from electrum.crypto import sha256
@@ -103,6 +104,7 @@ class Jade_Client(HardwareClientBase):
         self.fwversion = _versiontuple(verinfo['JADE_VERSION'])
         self.efusemac = verinfo['EFUSEMAC']
         self.jade.disconnect()
+        time.sleep(2)
 
         # Reconnecting with default timeout (for all subsequent calls)
         self.jade = JadeAPI.create_serial(device)
```

JamieDriver commented:
Right, I had got to a similar point.  ;-)
I spent some time this morning getting a win11 VM up and running - and I can now get the same behaviour you report. 
I also tested with green-qt (thanks for the relevant env var!) - and noticed that while Jade operates well, when I close the qt app, Jade reboots. This is not expected/usual.

The rub seems to be, when we drop the serial connection, Jade reboots.  
Your change above works because the sleep gives the unit time to restart, so the next 'connect()' reaches a newly rebooted unit.
However, tbh I don't think that 'connect-disconnect-reconnect' strategy is really necessary, so my latest push tried it just connecting once, and it appeared to work - so I'll quieten the logging (I had bumped up in the hope of getting more info) and re-push that shortly.

However, I am not 100% happy as I don't know exactly what is causing Jade to reboot on disconnect in our VMs.  It doesn't happen under the other OSes that I have seen/heard about - and indeed our tests (with real/proper Win11 host machines) do not exhibit this behaviour either.

The fact that it is happening also when the green_qt app disconnects (C++/qt app) makes me suspect some low-level lib or driver - or indeed something in the VM usb-passthrough ... (or did you say you also had the same issue on a 'real' windows machine ?)

As is often the case, this issue lies somewhere between 'interesting' and 'exasperating' !  ;-)

JamieDriver commented:
OK, let's see how this fares!   Cheers.

SomberNight commented:
>  I spent some time this morning getting a win11 VM up and running - and I can now get the same behaviour you report.

Ah, I am glad you have proof that it is not just me going crazy. :)

> The rub seems to be, when we drop the serial connection, Jade reboots.
> However, tbh I don't think that 'connect-disconnect-reconnect' strategy is really necessary, so my latest push tried it just connecting once, and it appeared to work

Right... although I sometimes see reboots in other cases too just not as reliably.
For example,
```python
jade = JadeAPI.create_serial(path)
jade.connect()
jade.get_version_info()
```
I sometimes get a reboot on the `get_version_info`, which will then block (also blocking the GUI in return) until the timeout - sometimes it happens five times in a row, and then I cannot reproduce it for minutes. o.O
Meaning even with the current PR, it sometimes does not work well for me - but at least it often does now :D

> The fact that it is happening also when the green_qt app disconnects (C++/qt app) makes me suspect some low-level lib or driver - or indeed something in the VM usb-passthrough ... (or did you say you also had the same issue on a 'real' windows machine ?)

Yes, I can reproduce the above described issues on a real Windows machine, running Win10 21H1.
It is not limited to VMs.

SomberNight commented on electrum/plugins/jade/jadepy/jade.py:
```diff
@@ -0,0 +1,656 @@
+# The default implementation used in JadeAPI._jadeRpc() below.
+# NOTE: Only available if the 'requests' dependency is available.
+try:
+    import requests
```
I've just realised that the PIN verification involves making HTTP network requests.
This is quite different from the other hwds where PIN verification happens on-device.
I assume this is intended. Do you have a high-level description somewhere how this works?

In any case, if the network requests are needed, the requests should respect the user's locally set proxy. Not sure how easy to do that is though; but could you take a look?

see e.g. `Network.send_http_on_proxy`:
https://github.com/spesmilo/electrum/blob/232e38e27dea91144ce4b2c057cc8db7de122936/electrum/network.py#L1320

JamieDriver commented:
So I also have some info re: the unexpected jade reboots on windows.
1. Newer Jade units use a serial chip CH9102, which works with windows 'out of the box' (windows [update] can locate a appropriate driver by itself) and does not need any additional drivers.  It does not cause Jade reboots and everything works as expected/desired - ie. same as linux.
2. Some older jade units use serial chip CP2104.  While this works well on linux, it may require an additional driver in [some versions of?] windows and macos.  Hence the link on the Blockstream site.

I believe all new jade units manufactured recently, and going forward, will use CH9102, and so hopefully this issue is restricted to a smaller number of older hw devices.
```
CHIP	VID	PID
CP2104	10C4	EA60	original and best!
CH9102	1A86	55D4	new and improved!
```
@SomberNight - not sure how you got sent one of the older ones as we've been on the new ones for a while now ... I guess you may have been shipped some from a dev office that's had them in a drawer for a while ;-)

JamieDriver commented:
Ok, I dug through the release notes for that driver, and they made reference to setting the RTS line.
Jade hw (serial chip -> esp32) has this wired as a 'reset' signal - added some code to the jade_serial.py to set that line to False - I wasn't sure this would work as the driver has plenty of opportunity to mess with this after any higher level code ... but it appears to work afaics!

If you agree the reboots have stopped, we're in a good place I think !

(In parallel I will push these changes to Jade repo master branch also.)

SomberNight commented:
> So I also have some info re: the unexpected jade reboots on windows.
> 1. Newer Jade units use a serial chip CH9102, which works with windows 'out of the box' [...]
> 2. Some older jade units use serial chip CP2104.  While this works well on linux, it may require an additional driver in [some versions of?] windows and macos. [...]

Ah, ok, there are different revisions, that explains our differing experiences. :)

> Ok, I dug through the release notes for that driver, and they made reference to setting the RTS line.
Jade hw (serial chip -> esp32) has this wired as a 'reset' signal - added some code to the jade_serial.py to set that line to False - I wasn't sure this would work as the driver has plenty of opportunity to mess with this after any higher level code ... but it appears to work afaics!
> 
> If you agree the reboots have stopped, we're in a good place I think !

It does not work for me. I get this exception:
```
read_cbor_message(). message=b'ownl'
E | plugin.DeviceMgr | failed to create client for jade at COM5: TypeError("a bytes-like object is required, not 'str'")
Traceback (most recent call last):
  File "...\electrum\electrum\plugin.py", line 601, in unpaired_device_infos
    client = self.create_client(device, handler, plugin)
  File "...\electrum\electrum\plugin.py", line 362, in wrapper
    return run_in_hwd_thread(partial(func, *args, **kwargs))
  File "...\electrum\electrum\plugin.py", line 355, in run_in_hwd_thread
    return fut.result()
  File "...\Python310\lib\concurrent\futures\_base.py", line 446, in result
    return self.__get_result()
  File "...\Python310\lib\concurrent\futures\_base.py", line 391, in __get_result
    raise self._exception
  File "...\Python310\lib\concurrent\futures\thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
  File "...\electrum\electrum\plugin.py", line 452, in create_client
    client = plugin.create_client(device, handler)
  File "...\electrum\electrum\plugin.py", line 362, in wrapper
    return run_in_hwd_thread(partial(func, *args, **kwargs))
  File "...\electrum\electrum\plugin.py", line 352, in run_in_hwd_thread
    return func()
  File "...\electrum\electrum\plugins\jade\jade.py", line 416, in create_client
    client = Jade_Client(device.path, plugin=self)
  File "...\electrum\electrum\plugins\jade\jade.py", line 113, in __init__
    verinfo = self.jade.get_version_info()
  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 185, in get_version_info
    return self._jadeRpc('get_version_info')
  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 159, in _jadeRpc
    reply = self.jade.make_rpc_call(request, long_timeout)
  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 654, in make_rpc_call
    reply = self.read_response(long_timeout)
  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 633, in read_response
    return self.read_cbor_message()
  File "...\electrum\electrum\plugins\jade\jadepy\jade.py", line 605, in read_cbor_message
    if 'id' in message:
TypeError: a bytes-like object is required, not 'str'
```

SomberNight commented on electrum/plugins/jade/jade.py:
```diff
@@ -0,0 +1,471 @@
+    return multisig_name
+
+# Helper to adapt Jade's http call/data to Network.send_http_on_proxy()
+def _http_request(params):
+    # Use the first non-onion url
+    url = [url for url in params['urls'] if not url.endswith('.onion')][0]
+    method = params['method'].lower()
+    json_payload = params.get('data')
+    json_response = Network.send_http_on_proxy(method, url, json=json_payload)
+    return {'body': json.loads(json_response)}
```
This looks good.

Could we also disable the fallback default `_http_request` implementation in jadepy, for sanity?
I don't really like this API where the http func has to be passed explicitly with every function call otherwise there is a silent fallback :P It has good usability but I would like to err on the explicit-failure-side of things here.

If we wanted to monkey-patch it from here, we could do e.g.
```python
try:
    from .jadepy import jade
except ImportError:
    pass
else:
    if hasattr(jade, "_http_request"):
        del jade._http_request
    del jade
```
But I guess it might be useful even for upstream to have some global flag in `jadepy` we can set to disable the default impl.

JamieDriver replied:
ok for now we can just comment the code entirely or delete the function as in your patch.
In parallel, I'll work the suggested improvements into the jade repo, and at some point we can refresh the jadepy here from a later tag/release of Jade.

SomberNight replied:
Sure. Thanks.

JamieDriver commented:
1. Commented default _http_request() - will work better solution into jade repo api.
2. Added check when deserialising cbor response (also adding to jade master)
3. Updated jadepy README.md, noting intent to improve jadepy, and refresh from a later tagged release.

So I only managed to reproduce your error once, and I think that was by 'claiming' the Jade usb device in windows guest, while there was ongoing communication with jade on linux host.  The change for (2) above meant we saw some logging, but it didn't help as at this point the jade had crashed and needed rebooting.
At that point it was fine.

Following the procedure you suggest here: https://github.com/spesmilo/electrum/pull/7633#issuecomment-1129087586 all seemed fine to me.  ( I too have an 'old' Jade to test with, and have that Silicon Labs driver installed in my win VM, so we *should* get similar behaviours ... )

JamieDriver commented:
I think this is as good as I can get it.  
The only time I get the unexepected jade reboot is if connecting and interacting with Jade on the linux host, then 'claim' that usb device in the windows guest, and run electrum there.
The first attempt to connect to Jade in Windows causes the reboot (which causes scan to hang until the timeout).

However, if I don't interact with Jade on the host, all seems good to me.
ie.
- physically connect/boot Jade device to windows machine
- run electrum
- scan, connect, all good

I can open and close the jade wallet on electrum, I can even unplug jade, and reconnect it - it will then ask for PIN again before any attempt to use it (sign, show address, etc).

All the time I stick with the windows machine, everything seems good.

JamieDriver commented:
I think we can prob just update the timeout on the existing connection - I'll have a dig.

JamieDriver commented:
As discussed.  
Now, even if Jade does reboot scanning fails pretty quickly and can be retried.
After reboot (or while rebooting) this may result in the 'Access Denied' error - but another retry and it finds the Jade, and then all seems well.

As discussed above, hopefully this only affects a small number of the older Jade units.

JamieDriver commented:
Many thanks for all your time and assistance on this - we are pretty excited about getting Jade working with Electrum wallet ;-)

SomberNight reviewed (approved):
Looks good. Thanks for your work and contribution. :)
